src/relay/pass/to_anf.cc

```diff
@@ -0,0 +1,474 @@
+
+  void VisitExpr_(const ConstantNode* c) final { }
+
+  void VisitExpr_(const OpNode* o) final { }
+};
+
+DependencyGraph DependencyGraph::Create(common::Arena* arena, const Expr& body) {
+  return Creator(arena).Create(body);
+}
+
+Expr ToANF(const Expr& e, const Module& m, std::set<GlobalVar>* gv);
+
+struct ScopeNode;
+using Scope = std::shared_ptr<ScopeNode>;
+
+/* Invariant: when parent is null level is 0
+ *
+ * Invariant: when parent is not null level is 1 + parent->level
+ */
+struct ScopeNode {
+  size_t level;
+  Scope parent;
+  std::shared_ptr<LetList> ll = std::make_shared<LetList>();
+  explicit ScopeNode(const Scope& parent) : level(1 + parent->level), parent(parent) { }
+  ScopeNode() : level(0) { }
+};
+
+Scope ChildScope(const Scope& s) {
+  return std::make_shared<ScopeNode>(s);
+}
+
+Scope LCA(Scope lhs, Scope rhs) {
+  while (lhs != rhs) {
+    if (lhs->level > rhs->level) {
+      lhs = lhs->parent;
+    } else if (lhs->level < rhs->level) {
+      rhs = rhs->parent;
+    } else {
+      lhs = lhs->parent;
+      rhs = rhs->parent;
+    }
+  }
+  return lhs;
+}
+
+using Edge = std::pair<Expr, Expr>;
+
+struct EdgeHash {
+  size_t operator()(const Edge& p) const {
+    return dmlc::HashCombine(NodeHash()(p.first), NodeHash()(p.second));
+  }
+};
+
+class ExprScopeMap {
+ public:
+  Scope GetScope(const Expr& e) const {
+    CHECK_NE(expr_scope_.count(e), 0);
+    return expr_scope_.at(e);
+  }
+
+  Scope GetScope(const Edge& edge) const {
+    return edge_scope_.count(edge) != 0 ? edge_scope_.at(edge) : GetScope(edge.first);
+  }
+
+  Scope GetScope(const Expr& e, const LinkedList<DependencyGraph::Node*>& es) const {
+    auto it = es.head;
+    if (it == nullptr) {
+      return global_scope_;
+    } else {
+      Scope s = GetScope(Edge(dg_.node_unmap.at(it->value), e));
+      it = it->next;
+      for (; it != nullptr; it = it->next) {
+        s = LCA(s, GetScope(Edge(dg_.node_unmap.at(it->value), e)));
+      }
+      return s;
+    }
+  }
+
+  // scoped_children should only contain the direct immediate children that has their own scope.
+  void DeclareSubScope(const Expr& e, const std::vector<Expr>& scoped_children) {
+    Scope s = GetScope(e);
+    std::vector<Scope> subscope;
+    for (const Expr& child : scoped_children) {
+      Scope ns = ChildScope(s);
+      subscope.push_back(ns);
+      Edge edge(e, child);
+      CHECK_EQ(edge_scope_.count(edge), 0);
+      edge_scope_.insert({edge, ns});
+    }
+    CHECK_EQ(expr_subscope_.count(e), 0);
+    expr_subscope_.insert({e, subscope});
+  }
+
+  void DeclareScope(const Expr& e, const Scope& s) {
+    CHECK_EQ(expr_scope_.count(e), 0);
+    expr_scope_.insert({e, s});
+  }
+
+  Scope GetSubScope(const Expr& e, size_t i) const {
+    return expr_subscope_.at(e).at(i);
+  }
+
+  explicit ExprScopeMap(const DependencyGraph& dg) : dg_(dg) { }
+
+ private:
+  // Scope of an expression.
+  std::unordered_map<Expr, Scope, NodeHash, NodeEqual> expr_scope_;
```

Comment:
We only need to assign scope to the Node* in the dependency graph, as long as we can map Expr to the Node* in the dependency graph it is fine